blog/2025-07-12-mtp.md

```diff
@@ -0,0 +1,120 @@
+* Data-Parallel Attention (DP Attention)
+* Expert Parallelism Load Balancer (EPLB)
+* DeepEP MoE
+* Two Batch Overlap
+* Prefill-Decode (PD) Disaggregation
+* CUDA Graph
+* Various Attention Backends
+
+## Performance Evaluation
+
+We present a comprehensive evaluation of the performance gains enabled by fully integrating MTP into the SGLang serving framework, using the DeepSeek V3 model as the testbed. This analysis includes two case studies designed to highlight improvements under both small-scale and large-scale deployment settings.
+
+### Deployment Scenarios and Design Motivation
+
+The small-scale deployment configuration was selected based on production requirements from a high-profile generative AI company that operates a latency-sensitive, developer-facing product. Specifically, the company required a minimum sustained output of 60.4 tokens/sec per rank to meet application-level service-level agreements (SLAs). This requirement guided the configuration used in our first case study. To assess scalability under heavier loads, we also evaluate MTP in a large-scale cluster setup.
+
+### Case Study 1: Small-Scale Deployment
+
+In this scenario, we deploy two decoding nodes across a total of 16 H200 GPUs, running 2 concurrent requests per rank with input sequence length of 65,536 tokens and output sequence length of 4,096 tokens.  As baseline, we tested the case with no MTP and no overlap scheduling, and the system achieves an output throughput of 51 tokens/sec per rank. Using overlap scheduling alone, a feature introduced in SGLang v0.4, we achieved 60.4 tokens/sec per rank, meeting the production threshold without the need for MTP. 
+When MTP is enabled, the system significantly surpasses this benchmark:
+* With a 3-token MTP window and topk=1, the system achieves a throughput of 81.5 tokens/sec per rank, with an average acceptance length of 2.18 tokens.
+*With a 4-token MTP window and topk=1, throughput increases to 82.0 tokens/sec per rank, with an average acceptance length of 2.44 tokens.
+
+![Small-scale throughput graph](/images/blog/mtp/small_scale_throughput_hr.png)
+
+These results represent a +60% improvement in output throughput compared to the baseline (i.e., no overlap scheduling and no MTP). This case demonstrates that MTP yields substantial performance gains even in smaller cluster settings with modest concurrency levels, allowing for scalable performance even within constrained GPU resource budgets.
+
+
+| Overlap Scheduling | MTP       | Throughput (tokens/sec) Per Rank |
+|--------------------|-----------|----------------------------------|
+| ❌                 | ❌        | 51.0 (baseline)                  |
+| ✅                 | ❌        | 60.4 (+20.4% ↑)                  |
+| ❌                 | ✅ 3-token| 81.5 (+59.8% ↑)                  |
+| ❌                 | ✅ 4-token| 82.0 (+60.8% ↑)                  |
+
+### Case Study 2: Large-Scale Deployment
```

Comment:
Explain the configuration of this. At the beginning, we highlighted we support MTP with EP and PD, we need to show an example.